File: 0042-trapping-rain-water/0042-trapping-rain-water.py
class Solution:
    def trap(self, height: List[int]) -> int:
        n = len(height)-1
        leftMax = height[0]
        rightMax = height[n]
        left = 0 
        right = n
        water = 0 
        while left<right:
            if leftMax<rightMax:
                water+=(leftMax-height[left])
                left+=1
                leftMax = max(leftMax, height[left])
            else:
                water+=(rightMax - height[right])
                right-=1
                rightMax = max(rightMax,height[right])
        return water
        

File: 0042-trapping-rain-water/0042-trapping-rain-water.cpp
class Solution {
public:
    int trap(vector<int>& height) {
        int left = 0 ;
        int right = height.size()-1;
        int leftMax = height[left];
        int rightMax = height[right];
        int water = 0 ;
        while(left<right)
        {
            if(leftMax<rightMax)
            {
                water+=leftMax - height[left];
                left++;
                leftMax = max(leftMax, height[left]);
            }
            else{
                water+=rightMax - height[right];
                right--;
                rightMax = max(rightMax, height[right]);
            }
        }

        return water;

        
    }
};